Handle User Bluetooth Disconnection Request

cvcore opened the pull request:
This PR addresses the issue that the keyboard will keep reconnecting itself to the computer, even if the user disconnects it manually from the bluetooth menu.

Related issue: https://github.com/KinesisCorporation/Adv360-Pro-ZMK/issues/172#issuecomment-1646550785

The previous behavior of zmk in the adv360-z3.2 branch will start open BLE advertising, when there exists no connection for the current profile, or if the current profile is open. This behavior will cause the keyboard to broadcast itself as connectable immediately after user BLE disconnection, causing the above mentioned behavior.

With the proposed change, the keyboard will no longer advertise itself as connectable after BLE disconnection. But wait until the user presses a key on the keyboard to retrigger the BLE advertising.

@ReFil please review and give suggestions.

petejohanson committed:
refactor(sensors): Sensor event channel data, resolution tweaks.

* Refactor sensor events to include channel data,
  necessary for prop split encoders, and avoiding duplicate calls,
  to fetch channel data twice, etc.
* More consistent behavior driver API.
* Allow setting triggers per resolution at the behavior level optionally.

petejohanson committed:
refactor(sensors): ec11 rotation sensor value in degrees.

* Add new `steps` property to the `aips,ec11` binding, to make
  the driver properly report degrees in the rotation delta channel.
* Handle old sensor values in sensor rotate behavior.

petejohanson committed:
feat(behaviors): Support parameterized macros.

* Add two new compatibles for macros that
  take one or two parameters when bound in
  a keymap.
* Use `&macro_param_1to1`, `&macro_param_1to2`, `&macro_param_2to1`,
  and `&macro_param_2to2` control entries in the bindings for the macro
  to have the next binding entry have it's values substituted.

Co-authored-by: Cem Aksoylar <caksoylar@users.noreply.github.com>

petejohanson committed:
refactor(config): Select SENSOR as needed.

* Don't force on SENSOR Kconfig setting unless
  needed based on the detected DT.

xudongzheng committed:
feat(split): allow central to connect to multiple peripherals

bortoz committed:
feat: handle LED indicators report

bortoz committed:
feat: LED indicators displaying

bortoz committed:
docs: add LED indicators documentation

bortoz committed:
feat: handle LED indicators report

ReFil committed:
feat(lighting): Kinesis lighting functionality

moved peripheral updating from behaviour bnased to dedicated bluetooth services, added custom lighting effects to the adv360

ReFil committed:
Battery life improvements

ReFil committed:
Change default settings

Dim brightness of LEDS for improved battery life, Decrease settings save timeout to 29.5 seconds so it occurs before idle timeout

hassa129 committed:
feat(split): filter known devices on scan

ReFil committed:
fix(lighting): fix initial power on lighting

ReFil committed:
Merge remote-tracking branch 'upstream/main' into adv360-z3

ReFil committed:
fix(bluetooth): Fix ble lights reporting

laserscout committed:
remove undeclared struct warning (#4)

ReFil committed:
chore(bluetooth): Fix clang formatting

ReFil committed:
Merge branch 'main' into adv360-z3

ReFil committed:
Fix profile 5 behaviour

petejohanson committed:
refactor(shields): Updated ZMK Uno encoder config.

* Move to new steps/triggers-per-rotation config.
* Leverage QDEC Nordic driver when used on Nordic DK.

petejohanson committed:
refactor(bluetooth): Add battery reporting config.

* Add dedicated battery reporting Kconfig that is `imply`d by
  enabling ZMK_BLE.

ReFil committed:
fix(underglow): fix RGB startup

ReFil committed:
fix(formatting) fix pre-commit formatting

ReFil committed:
Merge branch 'main' into adv360-z3.2

petejohanson committed:
refactor(sensors): Split data handling from triggers.

* All sensor behaviors should see sensor data, then selectively
  only have some trigger their behaviors.

joelspadin committed:
fix: Enable BT_TINYCRYPT_ECC when using HCI

BT_TINYCRYPT_ECC is required for BT_SMP_SC_PAIR_ONLY to work on setups
that use BT HCI like the nRF5340.

joelspadin committed:
refactor: Split endpoint to transport and instance

Changed the endpoints code to rename the existing endpoint types to
"endpoint transport" and add the concept of "endpoint instances". A
transport is the method by which data is sent, while instances allow
describing multiple endpoints that use the same transport (e.g.
bluetooth profiles).

Also added new APIs to get the total number of possible endpoint
instances and assign each instance a unique index, which can be used
for tracking separate state for each endpoint in other code files.

joelspadin committed:
fix(boards): Bump nRF5340 DK I2C buffer size

Increased the I2C buffer size again, since it needs to be at least 641
to support 128x64 displays.

petejohanson committed:
refactor(sensors): Use "sensor index" consistently

xudongzheng committed:
refactor(split): allow central to define connection parameters

Fixes #1614

petejohanson committed:
refactor(bluetooth): Bump HoG stack size.

* Bump the default stack size for the HoG processing thread
  to avoid issues w/ some pathways.

petejohanson committed:
feat(blog): Add post about sensor refactor.

* Document changed configuration, support for other Zephyr
  drivers, next steps.

Co-authored-by: Cem Aksoylar <caksoylar@users.noreply.github.com>

caksoylar committed:
feat(docs): Add note on modifier functions to macros

caksoylar committed:
feat(docs): Add behavior types section for hold-tap

caksoylar committed:
feat(docs): Add troubleshooting for Windows issue

caksoylar committed:
feat(docs): Add pointer to Bluetooth page in troubleshooting

ReFil committed:
Merge branch 'main' into adv360-z3.2

ReFil committed:
fix(zephyr): update new includes

ReFil committed:
refactor(ble): Match new kconfigs

ReFil committed:
fix(formatting): whitespace

ReFil committed:
refactor(battery): disable BLE battery reporting

Disable battery reporting, stops spurious wakeups on Macs

ReFil committed:
refactor(ble): Match new kconfigs

ReFil committed:
Fix(leds): Update indicators code

caksoylar committed:
refactor(docs): Move BT troubleshooting items to BT page

caksoylar committed:
feat(docs): Note split connectivity improvement with TX power

petejohanson committed:
fix(bluetooth): Passkey pairing improvements.

* Capture the last 6 entered digits, and then require pressing
   Enter/Return to submit the entered digits. This matches the
   messaging shown on hosts regarding how to complete pairing.
* Fix the wording on the Kconfig menu item to accurately describe
  the feature.

caksoylar committed:
fix(docs): Fix broken link in BT troubleshooting

petejohanson committed:
fix(boards): Proper i2c pinctrl for BlueMicro840

* Use the proper pin assignmets after the move to pinctrl for the
  Zephyr 3.2 migration.

filterpaper committed:
feat(docs): Add an example for combining just modifiers (#1826)

Co-authored-by: Cem Aksoylar <caksoylar@users.noreply.github.com>

kadoyau committed:
fix(docs): Fix INT6 keycode description

dixls committed:
fix(docs): Change user-setup.md order to agree with order in setup script

Fixes #1281 

Co-authored-by: Cem Aksoylar <caksoylar@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump minimatch and serve-handler in /docs

Bumps [minimatch](https://github.com/isaacs/minimatch) and [serve-handler](https://github.com/zeit/serve-handler). These dependencies needed to be updated together.

Updates `minimatch` from 3.0.4 to 3.1.2
- [Release notes](https://github.com/isaacs/minimatch/releases)
- [Commits](https://github.com/isaacs/minimatch/compare/v3.0.4...v3.1.2)

Updates `serve-handler` from 6.1.3 to 6.1.5
- [Release notes](https://github.com/zeit/serve-handler/releases)
- [Commits](https://github.com/zeit/serve-handler/compare/6.1.3...6.1.5)

---
updated-dependencies:
- dependency-name: minimatch
  dependency-type: indirect
- dependency-name: serve-handler
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
chore(deps-dev): bump webpack from 5.80.0 to 5.86.0 in /docs

Bumps [webpack](https://github.com/webpack/webpack) from 5.80.0 to 5.86.0.
- [Release notes](https://github.com/webpack/webpack/releases)
- [Commits](https://github.com/webpack/webpack/compare/v5.80.0...v5.86.0)

---
updated-dependencies:
- dependency-name: webpack
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

petejohanson committed:
fix(sensors): Clean ups based on code review.

petejohanson committed:
fix(sensors): Sensor rotate behavior fix for layers.

* Properly return transparent value when processing mode
  for higher disabled layers means the processing is ignored.

kylemccreery committed:
fix(boards): Update pillbug_defconfig to enable pinctrl

filterpaper committed:
feat(docs): Note that UF2 drive unmounts after flashing (#1859)

Add a note about the UF2 drive that automatically unmounts itself
to restart, which may surprise some new users.

petejohanson committed:
feat(shields): Add splitkb.com Aurora Sofle

* Add new shield for splitkb.com Aurora Sofle, supporting
  keys, encoder(s), displays, and RGB.

Co-authored-by: Cem Aksoylar <caksoylar@users.noreply.github.com>

dependabot[bot] committed:
chore(deps): bump semver from 5.7.1 to 5.7.2 in /docs

Bumps [semver](https://github.com/npm/node-semver) from 5.7.1 to 5.7.2.
- [Release notes](https://github.com/npm/node-semver/releases)
- [Changelog](https://github.com/npm/node-semver/blob/v5.7.2/CHANGELOG.md)
- [Commits](https://github.com/npm/node-semver/compare/v5.7.1...v5.7.2)

---
updated-dependencies:
- dependency-name: semver
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

xudongzheng committed:
refactor: use low priority workqueue for underglow and battery reporting

Blocking operations on the high priority system workqueue may result in
deadlocks, particularly when Bluetooth is in use.

ClicketySplit committed:
feat(boards): Leeloo v2 and V2 Zephyr 3.2 fixes

* Add Leeloo v2
* Leeloo-Micro v1 updates to support Zephyr 3.2
* Refactored for new Encoder Configuration

Co-authored-by: Cem Aksoylar <caksoylar@users.noreply.github.com>

xudongzheng committed:
refactor(underglow): turn underglow off in low priority work queue

trueNAHO committed:
fix(docs): Fix README punctuation and add license link

Nicell committed:
feat(nice!view): Custom widgets

Nicell committed:
fix(displays): Don't enable built-in widget configs by default

caksoylar committed:
fix: Proper battery sensor Kconfig dependencies.

Properly make the battery sensor Kconfig symbols depend
on `SENSOR` config, and minor battery reporting Kconfig
symbol dependency fix.

petejohanson committed:
fix(bluetooth): Properly clear peripheral slots

* When the clear bonds Kconfig is set, also clear peripheral
  address slots addresses from settings as well.

petejohanson committed:
fix(bluetooth): Corrected use of `bt_addr_le_cmp`

* Properly compare to zero when comparing LE addresses.

narfdotpl committed:
fix(docs): Fix typos in parameterized macros docs.

petejohanson committed:
feat: Add more logging to peripheral settings.

petejohanson committed:
fix(bluetooth): Split improvements

* Proper usage of bt_uuid_cmp.
* Central's don't start scanning for peripherals if
  `ZMK_BLE_CLEAR_BONDS_ON_START` is enabled.
* Split peripherals don't advertize if
  `ZMK_BLE_CLEAR_BONDS_ON_START` is enabled.

petejohanson committed:
fix(shields): Make settings_reset more flexible.

* Don't reference `pro_micro` nexus node in settings_reset
  so it can be used with other controllers.
* Use mock kscan node instead.

ReFil commented on app/src/ble.c:
```diff
@@ -48,7 +50,7 @@ static uint8_t passkey_digit = 0;
 
 #if IS_ENABLED(CONFIG_ZMK_SPLIT_ROLE_CENTRAL)
-#define PROFILE_COUNT (CONFIG_BT_MAX_PAIRED - 1)
+#define PROFILE_COUNT (CONFIG_BT_MAX_PAIRED - 1) // should it be ZMK_BLE_PROFILE_COUNT?
```
They are functionally the same, one pairing slot is reserved for the right half

cvcore replied:
Yes, but I can't find where `PROFILE_COUNT` was used anywhere in the code. Seems everywhere we are referring to `ZMK_BLE_PROFILE_COUNT` which made me a bit confused.

ReFil replied:
The number of profiles configured in ZMK is not the same as the technical limitation on the number of profiles. The PROFILE_COUNT value is defined as compile time based on the max pairing slots configured, you can have 8 pairing slots and then use ZMK_BLE_PROFILE_COUNT to limit the number of actual profiles

ReFil commented on app/src/ble.c:
```diff
@@ -63,12 +65,14 @@ enum advertising_type {
+bool active_profile_seeking_connection = true;
 
-#define DEVICE_NAME CONFIG_BT_DEVICE_NAME
+#define DEVICE_NAME CONFIG_BT_DEVICE_NAME // where is this defined?
```
anything prefixed with CONFIG_ gets set in one of the KConfig files, so adv360_*side*_defconfig as well as a few other files in the adv360 directory, a lot of parameters are set at compile time by the kconfig file in the root of app as well

ReFil commented on app/src/ble.c:
```diff
@@ -141,6 +146,7 @@ bool zmk_ble_active_profile_is_connected() {
         return err;                                                                                \
     }
 
+/* Start direct advertising to peer if no connection exists for active profile */
```
I believe direct advertising to hosts is currently not a thing because apple do not permit it to any apple device

ReFil commented:
It all looks good, but I'd like to se it behind a Kconfig option so as not to introduce a potentially confusing change to regulat users

ReFil committed:
fix(ble): Re enable battery report

cvcore commented:
@ReFil Thank you for the review and clarifying open questions! Sure, adding a default off switch is definitely better idea before it's introduced to everyone as a default.
I will make the adjustments in subsequent commits and let you know when it's ready for merging.

cvcore commented:
@ReFil Added the config with switch `CONFIG_ZMK_HANDLE_BLE_DISCONNECTION`, please have another review. Thanks!

ReFil commented:
You need to add the config option in the root KConfig otherwise the build system doesn't know whaat to do with it.

cvcore commented:
> You need to add the config option in the root KConfig otherwise the build system doesn't know whaat to do with it.

Good catch! Fixed 👍

ReFil commented:
Can you remove the extra comments you've added in some of the source files to streamline the changes please?

cvcore committed:
feat(ble): handle BLE disconnection request

- added switch `CONFIG_ZMK_HANDLE_BLE_DISCONNECTION`
- fix KinesisCorporation/Adv360-Pro-ZMK#172

cvcore commented:
Done :) lmk if any further change is needed.

ReFil commented:
Please add back the line ending in app/KConfig that got removed

DerDreschner committed:
Fix broken CI builds on Zephyr 3.2 (#6)

* fix(backlight/tests): Restore original behaviour to fix backlight tests

* fix(build): Remove feature-specific code when SPLIT_BLE, RGB_UNDERGLOW or BACKLIGHT isn't enabled

* refactor(bluetooth): Add battery reporting config.

* Add dedicated battery reporting Kconfig that is `imply`d by
  enabling ZMK_BLE.

* fix(rgb_underglow): Use correct condition in service.c

* Revert "refactor(bluetooth): Add battery reporting config."

This reverts commit 881da25a5affe141007f2c9fa938fc9d7f5753e5.

---------

Co-authored-by: Peter Johanson <peter@peterjohanson.com>

cvcore commented:
> Please add back the line ending in app/KConfig that got removed

Hmm interesting. I've reverted the change assuming you would prefer a cleaner git history. Apart from that is there a reason for having it there? Since most modern editors will remove trailing whitespaces by default. Just asking to avoid future confusions.

Actually I didn't notice the edit in the beginning since it's done automatically from VSCode, and upstream [Kconfig](https://github.com/zmkfirmware/zmk/blob/main/app/Kconfig) also doesn't have the trailing line.

ReFil commented:
Thank you, I didn't know about the upstream change. I think the cleaner change in this PR the better to try and minimise the number of merge conflicts I'll have to deal with when upstream is merged again. I'll test this PR soon and if all is good I'll merge it in

ReFil committed:
fixing changes

cvcore committed:
Merge branch 'adv360-z3.2' into adv360-z3.2-beta

cvcore commented:
FYI @ReFil : now updated the code base to `adv360-z3.2-beta` and tested on my hardware.

cvcore commented:
Closed for #9
